src/hotspot/share/opto/phaseX.cpp

```diff
@@ -1991,24 +1991,35 @@ void PhaseCCP::push_load_barrier(Unique_Node_List& worklist, const BarrierSetC2*
   }
 }
 
-// AndI/L::Value() optimizes patterns similar to (v << 2) & 3 to zero if they are bitwise disjoint.
-// Add the AndI/L nodes back to the worklist to re-apply Value() in case the shift value changed.
-// Pattern: parent -> LShift (use) -> (ConstraintCast | ConvI2L)* -> And
+// AndI/L::Value() optimizes patterns similar to (v << 2) & 3, or CON & 3 to zero if they are bitwise disjoint.
+// Add the AndI/L nodes back to the worklist to re-apply Value() in case the value is now a constant or shift
+// value changed.
 void PhaseCCP::push_and(Unique_Node_List& worklist, const Node* parent, const Node* use) const {
+  const Node* to_push = nullptr;
+
+  const TypeInteger* new_type = type(parent)->isa_integer(type(parent)->basic_type());
   uint use_op = use->Opcode();
-  if ((use_op == Op_LShiftI || use_op == Op_LShiftL)
-      && use->in(2) == parent) { // is shift value (right-hand side of LShift)
-    auto push_and_uses_to_worklist = [&](Node* n){
-      uint opc = n->Opcode();
-      if (opc == Op_AndI || opc == Op_AndL) {
-        push_if_not_bottom_type(worklist, n);
-      }
-    };
-    auto is_boundary = [](Node* n) {
-      return !(n->is_ConstraintCast() || n->Opcode() == Op_ConvI2L);
-    };
-    use->visit_uses(push_and_uses_to_worklist, is_boundary);
+  if (new_type != nullptr && new_type->is_con()) {
+    // Pattern: parent (now constant) -> (ConstraintCast | ConvI2L)* -> And
+    to_push = parent;
```

Comment:
Why do we need this?
Do we not already push all uses of `parent` further up?
```
  1881 void PhaseCCP::push_child_nodes_to_worklist(Unique_Node_List& worklist, Node* n) const {
  1882   for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
  1883     Node* use = n->fast_out(i);
  1884     push_if_not_bottom_type(worklist, use);
  1885     push_more_uses(worklist, n, use);                                                                                                                                                                                          
  1886   }
  1887 }
```

Reply:
What am I missing?

Reply:
Ah, that would not traverse through `ContraintCast` and `ConvI2L`. I see.

Reply:
Ah, why not just add these two cases to the case below?
`((use_op == Op_LShiftI || use_op == Op_LShiftL) && use->in(2) == parent`
or
`n->is_ConstraintCast() || n->Opcode() == Op_ConvI2L` ?

That would allow you to traverse this pattern too:
`// Pattern: parent (now constant) -> (ConstraintCast | ConvI2L)* -> And
`

I think this would be closer to existing patterns.

What do you think?